### PR TITLE
SecOps - 41204 - Interactive and long-term retention duration meets the security baseline on every Microsoft Sentinel workspace

### PR DESCRIPTION
**Pull request by praneeth-0000**

### CODE REVIEW

File: src/powershell/tests/Test-Assessment.41204.ps1 [hunk 1]
```diff
@@ -0,0 +1,378 @@
...
+    Pillar: SecOps
+    Category: Security information and event management
+    Required API: Azure Resource Manager (management.azure.com)
+#>
+function Test-Assessment-41204 {
+    [ZtTest(
+        Category = 'Security information and event management',
+        ImplementationCost = 'Low',
+        MinimumLicense = ('Consumption-based: Microsoft Sentinel'),
+        Pillar = 'SecOps',
```
[the remark this answers is not in the document]
Comment (praneeth-0000) [reply src/powershell/tests/Test-Assessment.41204.ps1:17]: We don't add compatible license when consumption based license is involved.

File: src/powershell/tests/Test-Assessment.41204.ps1 [hunk 2]
```diff
@@ -0,0 +1,378 @@
...
+    }
+
+    $failedItems      = @($workspaceResults | Where-Object { $_.RowStatus -eq 'Fail' })
+    $investigateItems = @($workspaceResults | Where-Object { $_.RowStatus -eq 'Investigate' })
+    $passed           = $failedItems.Count -eq 0 -and $investigateItems.Count -eq 0
+    $customStatus     = $null
+
+    if ($failedItems.Count -gt 0) {
+        $testResultMarkdown = "❌ The workspace default analytics retention is below 90 days, or a present security-relevant table has less than 365 days of total retention.`n`n%TestResult%"
+    }
```
[the remark this answers is not in the document]
Comment (praneeth-0000) [reply src/powershell/tests/Test-Assessment.41204.ps1:313]: The msg is according to the spec, not needed.

File: src/powershell/tests/Test-Assessment.41204.ps1 [hunk 3]
```diff
@@ -0,0 +1,378 @@
...
+                else {
+                    try {
+                        $tableRetentionDays = [int]$table.properties.retentionInDays
+                        $totalRetentionDays = [int]$table.properties.totalRetentionInDays
+                        $archiveDays = if ($null -ne $table.properties.archiveRetentionInDays) {
+                            [int]$table.properties.archiveRetentionInDays
+                        }
+                        else {
+                            $totalRetentionDays - $tableRetentionDays
+                        }
```
[the remark this answers is not in the document]
Comment (praneeth-0000) [reply src/powershell/tests/Test-Assessment.41204.ps1:233]: Azure enforces total retention to be greater than or equal to analytics retention. The portal does not allow analytics retention to exceed total retention, and the API returns the effective validated retention configuration.

File: src/powershell/tests/Test-Assessment.41204.ps1 [hunk 4]
```diff
@@ -0,0 +1,378 @@
...
+        $params = @{
+            TestId       = '41204'
+            Title        = 'Interactive and long-term retention duration meets the security baseline on every Microsoft Sentinel workspace'
+            Status       = $false
+            Result       = '⚠️ A workspace-discovery, onboarding, Q1, or Q2 call returned an authorization failure, throttling, service error, or malformed response; or a Sentinel-onboarded workspace exists but none of the security-relevant tables is present to evaluate.'
+            CustomStatus = 'Investigate'
+        }
+        Add-ZtTestResultDetail @params
+        return
+    }
```
[the remark this answers is not in the document]
Comment (praneeth-0000) [reply src/powershell/tests/Test-Assessment.41204.ps1]: fixed

File: src/powershell/tests/Test-Assessment.41204.ps1 [hunk 5]
```diff
@@ -0,0 +1,412 @@
...
+
+    Write-PSFMessage '🟦 Start' -Tag Test -Level VeryVerbose
+    $activity = 'Checking retention settings in Microsoft Sentinel workspaces'
+
+    # Q1 + Q2 + onboarding check via shared helper.
+    # Returns 'Forbidden'        on ARG 401/403 (Investigate).
+    # Returns $null              on unexpected ARG failure (Investigate).
+    # Returns 'NoSubscriptions'  when no enabled subscriptions are accessible (Skip).
+    # Returns 'NoWorkspaces'     when no Log Analytics workspaces exist in scope (Skip).
+    $allWorkspaces = Get-SentinelWorkspaceData -Activity $activity
```
Comment (alexandair) [inline src/powershell/tests/Test-Assessment.41204.ps1:39]: **A malformed onboarding response can be treated as onboarded and ultimately Pass**

41204 delegates discovery and onboarding entirely to `Get-SentinelWorkspaceData` at `Test-Assessment.41204.ps1#L30-L36`, then trusts its `SentinelOnboarded` flag at `Test-Assessment.41204.ps1#L78-L81`. The shared helper sets that flag for every HTTP 200 without parsing or validating the response body:

```powershell
switch ([int]$response.StatusCode) {
    200 { $sentinelOnboarded = $true }
```

Affected helper: `Get-SentinelWorkspaceData.ps1#L125-L140`.

The specification explicitly requires a malformed D2 body to mark that workspace `Investigate`, not onboarded. Microsoft Learn documents the onboarding-state resource body with a required name and a `properties` object containing `customerManagedKey`; status alone is not the complete response contract.

A focused mocked execution returned HTTP 200 with body `{}`, followed by otherwise passing Q1/Q2 values. The assessment completed with `Status = True` and no `CustomStatus`, reproducing a false Pass.

**Required change:** make the shared helper validate a successful onboarding response body before setting `SentinelOnboarded = $true`. A missing, unparseable, or structurally invalid body must set `OnboardingError = $true`, allowing 41204 to produce `Investigate`. Add a regression test for HTTP 200 with empty, invalid JSON, and structurally incomplete bodies.

File: src/powershell/tests/Test-Assessment.41204.ps1 [hunk 6]
```diff
@@ -0,0 +1,412 @@
...
+                    continue
+                }
+
+                $table              = $tablesByName[$tableName]
+                $tableRetentionDays = $null
+                $totalRetentionDays = $null
+                $archiveDays        = $null
+                $tableStatus        = 'Investigate'
+                $tableDetails       = ''
+
```
Comment (alexandair) [inline src/powershell/tests/Test-Assessment.41204.ps1:220]: **Internally inconsistent retention values are accepted as Pass**

The Q2 evaluator checks that `retentionInDays` and `totalRetentionInDays` exist and can be cast to integers, but it does not validate their semantic relationship. When `archiveRetentionInDays` is absent, it derives archive retention and then bases the verdict only on total retention:

```powershell
$archiveDays = if ($null -ne $table.properties.archiveRetentionInDays) {
    [int]$table.properties.archiveRetentionInDays
}
else {
    $totalRetentionDays - $tableRetentionDays
}
$tableStatus = if ($totalRetentionDays -ge 365) { 'Pass' } else { 'Fail' }
```

Affected code: `Test-Assessment.41204.ps1#L220-L258`.

Microsoft Learn defines total retention as including interactive retention and archive retention as `totalRetentionInDays - retentionInDays`. Therefore, total retention cannot be less than interactive retention in a valid response. The specification requires malformed Q2 data to become `Investigate`.

A focused execution with `retentionInDays = 400`, `totalRetentionInDays = 365`, and no archive field completed with `Status = True` and rendered `-35 days` of archive retention. This is both a false Pass and visibly impossible output.

**Required change:** validate retention invariants before threshold evaluation. At minimum, require non-negative values and `totalRetentionInDays >= retentionInDays`; when archive is supplied, validate that it is non-negative and consistent with the documented total. Violations must set the table to `Investigate`, not Pass or Fail. Add regression tests for negative, inverted, and inconsistent retention values.

File: src/powershell/tests/Test-Assessment.41204.ps1 [hunk 7]
```diff
@@ -0,0 +1,412 @@
...
+        $testResultMarkdown = "⚠️ A workspace-discovery, onboarding, Q1, or Q2 call returned an authorization failure, throttling, service error, or malformed response; or a Sentinel-onboarded workspace exists but none of the security-relevant tables is present to evaluate.`n`n%TestResult%"
+    }
+    else {
+        $testResultMarkdown = "✅ The workspace default analytics retention is at least 90 days, and every present security-relevant table has at least 365 days of total retention.`n`n%TestResult%"
+    }
+
+    #endregion Assessment Logic
+
+    #region Report Generation
+    $portalSentinelLink = "https://portal.azure.com/#browse/microsoft.securityinsightsarg%2Fsentinel"
```
Comment (alexandair) [inline src/powershell/tests/Test-Assessment.41204.ps1]: **All generated portal links are invalid for the supported USGov cloud**

The specification supports Global and USGov, but the report hardcodes `https://portal.azure.com` for the Sentinel overview and every subscription/workspace link at `Test-Assessment.41204.ps1#L325-L372`. Microsoft Learn identifies `https://portal.azure.us` as the Azure Government portal. Neighboring SecOps tests already select the host from `(Get-AzContext).Environment.Name`.

**Recommendation:** use the established `portalHost` pattern so AzureUSGovernment reports target `https://portal.azure.us` and commercial reports continue to use `https://portal.azure.com`.

**Review by alexandair — CHANGES_REQUESTED** [review]
@praneeth-0000 Please, address my feedback.

File: src/powershell/tests/Test-Assessment.41204.ps1 (hunk 7 above)
Comment (praneeth-0000) [reply src/powershell/tests/Test-Assessment.41204.ps1]: This is resolved.

File: src/powershell/tests/Test-Assessment.41204.ps1 (hunk 5 above)
Comment (praneeth-0000) [reply src/powershell/tests/Test-Assessment.41204.ps1:39]: Closing this comment as discussed, malformed is removed from spec.

File: src/powershell/tests/Test-Assessment.41204.ps1 (hunk 6 above)
Comment (praneeth-0000) [reply src/powershell/tests/Test-Assessment.41204.ps1:220]: Closing this comment as discussed, malformed is removed from spec.

**Review by alexandair — APPROVED** [review]
LGTM
